Common/Utils/include/CommonUtils/BoostHistogramUtils.h

```diff
@@ -0,0 +1,243 @@
+        AxisIterator operator++(int) {
+            AxisIterator result(mBaseIterator);
+            mBaseIterator++;
+            return result;
+        }
+
+        decltype(auto) operator*() {return mBaseIterator->center(); }
+    private:
+        AxisIterator    mBaseIterator;
+};
+
+template<typename AxisIterator>
+class BinUpperView {
+    public:
+        BinCenterView(AxisIterator iter)
+        AxisIterator& operator++() {
+            ++mBaseIterator;
+            return mBaseIterator;
+        }
+        AxisIterator operator++(int) {
+            AxisIterator result(mBaseIterator);
+            mBaseIterator++;
+            return result;
+        }
+
+        decltype(auto) operator*() {return mBaseIterator->upper(); }
+    private:
+        AxisIterator    mBaseIterator;
+};
+
+template<typename AxisIterator>
+class BinLowerView {
+    public:
+        BinCenterView(AxisIterator iter)
+        AxisIterator& operator++() {
+            ++mBaseIterator;
+            return mBaseIterator;
+        }
+        AxisIterator operator++(int) {
+            AxisIterator result(mBaseIterator);
+            mBaseIterator++;
+            return result;
+        }
+
+        decltype(auto) operator*() {return mBaseIterator->lower(); }
+    private:
+        AxisIterator    mBaseIterator;
+};
+
+/// fast fit of an array with ranges (histogram) with gaussian function
+///
+/// Fitting procedure:
+/// 1. Step - make logarithm
+/// 2. Linear  fit (parabola) - more robust, always converges, fast
+///
+/// \return chi2 or exit code
+///          >0: the chi2 returned by TLinearFitter
+///          -3: only three points have been used for the calculation - no fitter was used
+///          -2: only two points have been used for the calculation - center of gravity was uesed for calculation
+///          -1: only one point has been used for the calculation - center of gravity was uesed for calculation
+///          -4: invalid result!!
+///
+template <typename Iterator, typename AxisIterator, typename T>
+double fitGaus(Iterator first, Iterator last, AxisIterator axisfirst,  std::vector<T>& param)
+{
+  static TLinearFitter fitter(3, "pol2");
+  static TMatrixD mat(3, 3);
+  static Double_t kTol = mat.GetTol();
+  fitter.StoreData(kFALSE);
+  fitter.ClearPoints();
+  TVectorD par(3);
+  TVectorD sigma(3);
+  TMatrixD A(3, 3);
+  TMatrixD b(3, 1);
+  T rms = TMath::RMS(nBins, arr);
+  T max = TMath::MaxElement(nBins, arr);
+  T binWidth = (xMax - xMin) / T(nBins);
+
+  Float_t meanCOG = 0;
+  Float_t rms2COG = 0;
+  Float_t sumCOG = 0;
+
+  Float_t entries = 0;
+  Int_t nfilled = 0;
+
+  param.resize(4);
+  param[0] = 0.;
+  param[1] = 0.;
+  param[2] = 0.;
+  param[3] = 0.;
+
+  for (Int_t i = 0; i < nBins; i++) {
+    entries += arr[i];
+    if (arr[i] > 0) {
+      nfilled++;
+    }
+  }
+
+  // TODO: Check why this is needed
+  if (max < 4) {
+    return -4;
+  }
+  if (entries < 12) {
+    return -4;
+  }
+
+  if (rms < kTol) {
+    return -4;
+  }
```

Comment:
Within the EMCAL library I would insist on handing errors as exception, and certainly not encoded in a parameter value. As this file is not part of EMCAL itself others have to decide.